base/message_loop/message_pump_io_starboard.cc

```diff
@@ -109,27 +107,83 @@ MessagePumpIOStarboard::~MessagePumpIOStarboard() {
   SbSocketWaiterDestroy(waiter_);
 }
 
+bool MessagePumpIOStarboard::UnregisterInterest(SbSocket socket,
+                                                int dropped_interests,
+                                                SocketWatcher* controller) {
+  DCHECK(SbSocketIsValid(socket));
+  DCHECK(controller);
+  DCHECK(dropped_interests == kSbSocketWaiterInterestRead ||
+         dropped_interests == kSbSocketWaiterInterestWrite ||
+         dropped_interests ==
+             (kSbSocketWaiterInterestRead | kSbSocketWaiterInterestWrite));
+  DCHECK_CALLED_ON_VALID_THREAD(watch_socket_caller_checker_);
+
+  // Make sure we don't pick up any funky internal masks.
+  int old_interest_mask =
+      controller->interests() &
+      (kSbSocketWaiterInterestRead | kSbSocketWaiterInterestWrite);
+  int interests = old_interest_mask & (~dropped_interests);
+  if (interests == old_interest_mask) {
+    // Interests didn't change, return.
+    return true;
+  }
+
+  SbSocket old_socket = controller->Release();
+  if (SbSocketIsValid(old_socket)) {
+    // It's illegal to use this function to listen on 2 separate fds with the
+    // same |controller|.
+    if (old_socket != socket) {
+      NOTREACHED() << "Sockets don't match" << old_socket << "!=" << socket;
+      return false;
+    }
+
+    // Must disarm the event before we can reuse it.
+    SbSocketWaiterRemove(waiter_, old_socket);
+  } else {
+    interests = kSbSocketWaiterInterestNone;
+  }
+  controller->set_interests(interests);
+
+  if (!SbSocketIsValid(socket)) {
+    NOTREACHED() << "Invalid socket" << socket;
+    return false;
+  }
+
+  if (interests) {
+    // Set current interest mask and waiter for this event.
+    if (!SbSocketWaiterAdd(waiter_, socket, controller,
+                           OnSocketWaiterNotification, interests,
+                           controller->persistent())) {
+      SB_DLOG(ERROR) << __FUNCTION__
+                     << " Failed to remain interested in interests="
+                     << interests;
+      return false;
+    }
+    controller->Init(socket, controller->persistent());
+  }
+  return true;
+}
+
 bool MessagePumpIOStarboard::Watch(SbSocket socket,
                                    bool persistent,
-                                   int mode,
+                                   int interests,
                                    SocketWatcher* controller,
                                    Watcher* delegate) {
   DCHECK(SbSocketIsValid(socket));
   DCHECK(controller);
   DCHECK(delegate);
-  DCHECK(mode == WATCH_READ || mode == WATCH_WRITE || mode == WATCH_READ_WRITE);
+  DCHECK(interests == kSbSocketWaiterInterestRead ||
+         interests == kSbSocketWaiterInterestWrite ||
+         interests ==
+             (kSbSocketWaiterInterestRead | kSbSocketWaiterInterestWrite));
+  if (interests == kSbSocketWaiterInterestNone) {
```

Comment:
Maybe we don't want to do that much work (which would be fine), but can't we do nothing anytime `controller->interests() & interests == interests` is true?

Reply:
Correct, changed (both cases are unlikely but theoretically possible).